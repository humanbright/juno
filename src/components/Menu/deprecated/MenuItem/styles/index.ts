export * from './StyledMenuItem';
export * from './StyledMuiListItemIcon';
export * from './StyledRcListItemLabel';
