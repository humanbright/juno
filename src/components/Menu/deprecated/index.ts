export * from './Menu';
export * from './MenuList';
export * from './MenuItem';
export * from './SubMenu';
