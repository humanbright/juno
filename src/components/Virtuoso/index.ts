export * from './Virtuoso';
export * from './utils';
