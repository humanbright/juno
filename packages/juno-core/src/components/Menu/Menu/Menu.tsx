import React, {
  ComponentProps,
  forwardRef,
  memo,
  useMemo,
  useState,
} from 'react';

import MuiMenu from '@material-ui/core/Menu';

import {
  combineClasses,
  RcBaseProps,
  styled,
  useEventCallback,
  useId,
  useRcPortalWindowContext,
  useThemeProps,
} from '../../../foundation';
import type {
  RcMenuContextType,
  RcMenuOnCloseReasonsType,
} from './MenuContext';
import { RcMenuContext } from './MenuContext';
import { MenuStyle } from './styles';
import { RcMenuClasses } from './utils';

type RcMenuProps = {
  /** auto close menu when menu item clicked */
  autoClose?: boolean;
  /** emit when menu closed */
  onClose?: (event: {}, reason: RcMenuOnCloseReasonsType) => void;
} & RcBaseProps<
  ComponentProps<typeof MuiMenu>,
  'disableEnforceFocus' | 'onClose'
>;

const MemoMuiMenu = memo(MuiMenu);

const _RcMenu = forwardRef<any, RcMenuProps>((inProps: RcMenuProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcMenu' });
  const { classes: classesProp, children, autoClose, onClose, ...rest } = props;
  const [focusedMenuItemId, setFocusedMenuItemId] = useState(null);
  const id = useId('menu-', true);

  const { externalWindow } = useRcPortalWindowContext();

  const classes = useMemo(
    () => combineClasses(RcMenuClasses, classesProp),
    [classesProp],
  );

  const handleClose = useEventCallback(
    (event: {}, reason: RcMenuOnCloseReasonsType) => {
      onClose?.(event, reason);
    },
  );

  const ctxValue: RcMenuContextType = useMemo(
    () => ({
      menuId: id,
      autoClose: !!autoClose,
      focusedMenuItemId,
      setFocusedMenuItemId,
      closeMenu: handleClose,
    }),
    [autoClose, focusedMenuItemId, handleClose, id],
  );

  return (
    <RcMenuContext.Provider value={ctxValue}>
      <MemoMuiMenu
        container={externalWindow?.document.body}
        {...rest}
        ref={ref}
        classes={classes}
        onClose={handleClose}
        disableEnforceFocus
      >
        {children}
      </MemoMuiMenu>
    </RcMenuContext.Provider>
  );
});

const RcMenu = styled(_RcMenu)`
  ${MenuStyle};
`;

RcMenu.defaultProps = {
  variant: 'selectedMenu',
  getContentAnchorEl: null,
  anchorReference: 'anchorEl',
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

RcMenu.displayName = 'RcMenu';

export { RcMenu, RcMenuContext };

export type { RcMenuContextType, RcMenuProps };
