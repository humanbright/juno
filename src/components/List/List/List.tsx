import MuiList from '@material-ui/core/List';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  RcBaseProps,
  styled,
  useThemeProps,
} from '../../../foundation';
import { ListStyle } from './styles';
import { RcListClasses } from './utils';

type RcListProps = {} & RcBaseProps<ComponentProps<typeof MuiList>>;

const _RcList = forwardRef<any, RcListProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcList' });
  const { classes: classesProp, children, ...rest } = props;
  const classes = useMemo(
    () => combineClasses(RcListClasses, classesProp),
    [classesProp],
  );

  return (
    <MuiList {...rest} ref={ref} classes={classes}>
      {children}
    </MuiList>
  );
});

const RcList = styled(_RcList)`
  ${ListStyle}
`;

RcList.defaultProps = {
  disablePadding: true,
};

RcList.displayName = 'RcList';

export { RcList };
export type { RcListProps };
