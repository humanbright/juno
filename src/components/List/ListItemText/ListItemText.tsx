import MuiListItemText from '@material-ui/core/ListItemText';
import React, { ComponentProps, forwardRef, useMemo } from 'react';

import {
  combineClasses,
  omit,
  RcBaseProps,
  RcPaletteProp,
  styled,
  withDeprecatedCheck,
  useThemeProps,
} from '../../../foundation';
import { RcTypographyProps } from '../../Typography';
import { ListItemTextStyle } from './styles';
import { RcListItemTextClasses } from './utils';

type TypographyProps = RcBaseProps<RcTypographyProps, 'weight'>;

type RcListItemTextProps = {
  /** is text Ellipsis, default is true. */
  isEllipsis?: boolean;
  /** These props will be forwarded to the primary typography component (as long as disableTypography is not true). */
  primaryTypographyProps?: TypographyProps;
  /** These props will be forwarded to the secondary typography component (as long as disableTypography is not true). */
  secondaryTypographyProps?: TypographyProps;
  /** @deprecated define color for primary, recommend use primaryTypographyProps color prop */
  primaryColor?: RcPaletteProp;
  /** @deprecated is alignCenter, recommend using classes to define it. */
  alignCenter?: boolean;
  /** @deprecated is lineThrough, recommend using classes to define it. */
  lineThrough?: boolean;
} & RcBaseProps<
  ComponentProps<typeof MuiListItemText>,
  'primaryTypographyProps' | 'secondaryTypographyProps'
>;

const _RcListItemText = forwardRef<any, RcListItemTextProps>((inProps, ref) => {
  const props = useThemeProps({ props: inProps, name: 'RcListItemText' });
  const {
    classes: classesProp,
    children,
    isEllipsis,
    primaryTypographyProps: primaryTypographyPropsProp,
    secondaryTypographyProps: secondaryTypographyPropsProp,
    primaryColor,
    alignCenter,
    lineThrough,
    ...rest
  } = props;

  const classes = useMemo(
    () => combineClasses(RcListItemTextClasses, classesProp),
    [classesProp],
  );

  const primaryTypographyProps = useMemo(
    () => omit(primaryTypographyPropsProp, ['color', 'variant']),
    [primaryTypographyPropsProp],
  );

  const secondaryTypographyProps = useMemo(
    () => omit(secondaryTypographyPropsProp, ['color', 'variant']),
    [secondaryTypographyPropsProp],
  );

  return (
    <MuiListItemText
      {...rest}
      ref={ref}
      classes={classes}
      primaryTypographyProps={primaryTypographyProps}
      secondaryTypographyProps={secondaryTypographyProps}
    >
      {children}
    </MuiListItemText>
  );
});

const RcListItemText = styled(
  withDeprecatedCheck(
    _RcListItemText,
    [
      {
        prop: 'primaryColor',
        time: '2021-4',
        comment: 'recommend use primaryTypographyProps color prop',
      },
      {
        prop: 'alignCenter',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
      {
        prop: 'lineThrough',
        time: '2021-4',
        comment: `recommend using classes to define it`,
      },
    ],
    'RcListItemText',
  ),
)`
  ${ListItemTextStyle}
`;

RcListItemText.defaultProps = {
  isEllipsis: true,
  primaryTypographyProps: {
    variant: 'body1',
    color: 'neutral.f06',
  },
  secondaryTypographyProps: {
    variant: 'caption1',
    color: 'neutral.f04',
  },
};

RcListItemText.displayName = 'RcListItemText';

export { RcListItemText };
export type { RcListItemTextProps };
