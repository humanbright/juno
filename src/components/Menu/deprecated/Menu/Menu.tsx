import MuiMenu from '@material-ui/core/Menu';
import React, {
  ComponentProps,
  forwardRef,
  memo,
  useMemo,
  useState,
} from 'react';

import {
  RcBaseProps,
  styled,
  useEventCallback,
  useRcPortalWindowContext,
  withDeprecatedCheck,
} from '../../../../foundation';
import { StyledMenu } from './styles';

type RcMenuProps = {
  autoClose?: boolean;
  /** @deprecated use autoClose to replace that */
  enableCloseMenuWhenItemClicked?: boolean;
} & RcBaseProps<ComponentProps<typeof MuiMenu>>;

type CloseReasonType = 'backdropClick' | 'escapeKeyDown' | 'tabKeyDown';

type MenuContextType = {
  autoClose: boolean;
  open: boolean;
  closeMenu: (event: {}, reason: CloseReasonType) => void;
  focusedMenuItemId: string;
  setFocusedMenuItemId: (id: string) => void;
};

const MenuContext = React.createContext<MenuContextType>({
  autoClose: false,
  open: false,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  closeMenu: (event: {}, reason: CloseReasonType) => {},
  focusedMenuItemId: '',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  setFocusedMenuItemId: (id: string) => {},
});

// TODO: should use ref to refactor this menu logic with focus item set
const MemoMenu = memo(StyledMenu);

const _RcMenu = forwardRef<any, RcMenuProps>(
  (
    {
      onClose,
      open,
      enableCloseMenuWhenItemClicked,
      autoClose = enableCloseMenuWhenItemClicked,
      children,
      ...rest
    },
    ref,
  ) => {
    const [focusedMenuItemId, setFocusedMenuItemId] = useState('');
    const { externalWindow } = useRcPortalWindowContext();

    const handleClose = useEventCallback(
      (event: {}, reason: CloseReasonType) => {
        onClose?.(event, reason as Exclude<CloseReasonType, 'tabKeyDown'>);
      },
    );

    const ctxValue: MenuContextType = useMemo(
      () => ({
        autoClose: !!autoClose,
        open,
        closeMenu: handleClose,
        focusedMenuItemId,
        setFocusedMenuItemId: (id: string) => setFocusedMenuItemId(id),
      }),
      [autoClose, focusedMenuItemId, handleClose, open],
    );

    return (
      <MenuContext.Provider value={ctxValue}>
        <MemoMenu
          container={externalWindow?.document.body}
          {...rest}
          open={open}
          ref={ref}
          onClose={handleClose}
          disableEnforceFocus
        >
          {children}
        </MemoMenu>
      </MenuContext.Provider>
    );
  },
);

/** @release */
const RcMenu = styled(
  withDeprecatedCheck(
    _RcMenu,
    [
      {
        prop: 'enableCloseMenuWhenItemClicked',
        time: '2021-2',
        comment: `use autoClose to replace that `,
      },
    ],
    'RcMenu',
  ),
)``;

RcMenu.displayName = 'RcMenu';

RcMenu.defaultProps = {
  getContentAnchorEl: null,
  anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
};

export { RcMenu, RcMenuProps, MenuContext as RcMenuContext };
