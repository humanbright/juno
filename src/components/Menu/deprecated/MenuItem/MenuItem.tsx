import { MenuItemProps as MuiMenuItemProps } from '@material-ui/core/MenuItem';
import clsx from 'clsx';
import uniqueId from 'lodash/uniqueId';
import React, {
  forwardRef,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import {
  Classes,
  combineClasses,
  omit,
  RcBaseProps,
  RcBaseSize,
  RcClassesProps,
  styled,
  useEventCallback,
  withDeprecatedCheck,
} from '../../../../foundation';
import checkSvg from '../../../../icon/Check';
import { RcIcon, RcIconProps } from '../../../Icon';
import { withTooltip } from '../../../Tooltip';
import { RcMenuContext } from '../Menu';
import { RcMenuListContext } from '../MenuList';
import { RcSubMenuContext } from '../SubMenu';
import {
  StyledMenuItem,
  StyledMuiListItemIcon,
  StyledRcIcon,
  StyledRcListItemLabel,
} from './styles';
import { RcMenuItemClasses } from './utils/MenuItemUtil';

type MenuItemSize = RcBaseSize<'large' | 'medium'>;

type RcMenuItemType = 'checked' | 'selected';

type RcMenuItemClassesType = RcClassesProps<
  Classes<MuiMenuItemProps> | 'checked' | 'unchecked'
>;

type RcMenuItemProps = {
  /** MenuItem size */
  size?: MenuItemSize;
  /** MenuItem with subtext */
  subtext?: ReactNode;

  /** checked for that should render checked icon */
  type?: RcMenuItemType;
  /** if type to be `checked`, the checked prop for that should render checked icon */
  checked?: boolean;

  /** @deprecated  MenuItem with icon, pls use RcListItemIcon */
  icon?: ReactNode;
  /** @deprecated pass to the icon, pls use RcIcon with symbol prop */
  symbol?: RcIconProps['symbol'];
  /** @deprecated MenuItem with avatar, pls use ListItemAvatar */
  avatar?: React.ReactElement;
  /** @deprecated */
  label?: string;

  /** @deprecated the maxWidth of MenuItem, recommend pass classes to define it. */
  maxWidth?: number;
  /**
   * @deprecated
   * when hover on MenuItem, is cursor in shape of pointer.
   * Instead recommend using classes to define it.
   */
  isMember?: boolean;
  /**  @deprecated automationId, can use 'data-test-automation-id' replace */
  automationId?: string;
} & RcMenuItemClassesType &
  RcBaseProps<
    MuiMenuItemProps,
    'disableGutters' | 'button' | 'classes' | 'title' | 'dense' | 'divider'
  >;

const _RcMenuItem = forwardRef<any, RcMenuItemProps>((props, ref) => {
  const {
    icon,
    subtext,
    type,
    checked,
    value,
    symbol,
    children,
    disabled,
    classes,
    isMember,
    avatar,
    automationId,
    maxWidth,
    size,
    label,
    className,
    onMouseEnter,
    onClick,
    ...rest
  } = props;

  const isCheckedType = type === 'checked';

  const iconElement = useMemo(() => {
    if (React.isValidElement(avatar)) {
      return avatar;
    }
    if (symbol) {
      return <RcIcon size="small" symbol={symbol} />;
    }
    if (React.isValidElement(icon)) {
      return icon;
    }
    if (typeof icon === 'string') {
      return <RcIcon size="small">{icon}</RcIcon>;
    }
    return null;
  }, [avatar, icon, symbol]);

  const renderSubText = useMemo(() => {
    if (!isCheckedType) {
      return subtext;
    }
    return checked ? (
      <StyledRcIcon
        size="large"
        symbol={checkSvg}
        color={['interactive', 'f01']}
      />
    ) : (
      subtext
    );
  }, [checked, isCheckedType, subtext]);

  const [menuItemUUID, setMenuItemUUID] = useState<string | undefined>(
    undefined,
  );

  const menuContext = useContext(RcMenuContext);
  const menuListContext = useContext(RcMenuListContext);
  const subMenuContext = useContext(RcSubMenuContext);

  useEffect(() => {
    if (!menuListContext.isMenuList) {
      menuListContext.setIsMenuList();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    setMenuItemUUID(uniqueId('rc-menu-item-'));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleMenuItemMouseEnter = useEventCallback(
    (e: React.MouseEvent<HTMLLIElement>) => {
      onMouseEnter && onMouseEnter(e);

      if (!menuItemUUID) {
        return;
      }
      if (menuListContext.isMenuList) {
        menuListContext.setFocusedMenuItemId(menuItemUUID);
      } else {
        menuContext.setFocusedMenuItemId(menuItemUUID);
      }
    },
  );

  const handleMenuItemClick = useEventCallback(
    (e: React.MouseEvent<HTMLLIElement>) => {
      onClick && onClick(e);
      if (menuContext.autoClose) {
        menuContext.closeMenu(e, 'tabKeyDown');
      }
      if (subMenuContext.autoClose && subMenuContext.closeSubMenu) {
        subMenuContext.closeSubMenu();
      }
    },
  );

  const groupClasses = useMemo(
    () => combineClasses(RcMenuItemClasses, classes),
    [classes],
  );

  const toClasses = useMemo(
    () => omit(groupClasses!, ['checked', 'unchecked']),
    [groupClasses],
  );

  return (
    <StyledMenuItem
      data-test-automation-id={automationId}
      disabled={disabled}
      isMember={isMember}
      data-disabled={disabled}
      maxWidth={maxWidth}
      ref={ref}
      size={size}
      value={value}
      {...rest}
      classes={toClasses}
      onMouseEnter={handleMenuItemMouseEnter}
      onClick={handleMenuItemClick}
      className={clsx(className, {
        [groupClasses!.checked]: isCheckedType && checked,
        [groupClasses!.unchecked]: isCheckedType && !checked,
      })}
    >
      {iconElement && (
        <StyledMuiListItemIcon>{iconElement}</StyledMuiListItemIcon>
      )}
      {label && <StyledRcListItemLabel>{label}&nbsp;</StyledRcListItemLabel>}
      {children}
      {renderSubText}
    </StyledMenuItem>
  );
});

/** @release */
const RcMenuItem = styled(
  withDeprecatedCheck(
    withTooltip(_RcMenuItem),
    [
      {
        prop: 'isMember',
        time: '2021-1',
        comment:
          'isMember used for when hover on MenuItem, is cursor in shape of pointer. Please instead recommend using classes to define it.',
      },
      {
        prop: 'maxWidth',
        time: '2021-1',
        comment: 'Please instead recommend using classes to define it.',
      },
      {
        prop: 'label',
        time: '2021-1',
        comment: 'Please add label component and add to children prop',
      },
    ],
    'RcMenuItem',
  ),
)``;

RcMenuItem.displayName = 'RcMenuItem';

RcMenuItem.defaultProps = {
  size: 'medium',
};

export { RcMenuItem, RcMenuItemProps, RcMenuItemType, RcMenuItemClassesType };
