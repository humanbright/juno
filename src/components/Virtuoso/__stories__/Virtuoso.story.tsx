import { Meta, Story } from '@storybook/react/types-6-0';
import range from 'lodash/range';
import React, { ComponentProps } from 'react';

import {
  notControlInDocTable,
  notShowInDocTable,
  sortInDocTable,
} from '../../../storybook';
import { RcListItemText } from '../../List/ListItemText';
import { RcMenuItem } from '../../Menu/MenuItem';
import { Virtuoso } from '../Virtuoso';

export default {
  title: 'ðŸš€ Cleanup Components/Virtuoso',
  component: Virtuoso,
  argTypes: {
    ...sortInDocTable<keyof VirtuosoProps>([]),
    ...notControlInDocTable<keyof VirtuosoProps>([]),
    ...notShowInDocTable<keyof VirtuosoProps>([]),
  },
} as Meta;

type VirtuosoProps = ComponentProps<typeof Virtuoso>;

const menus = range(0, 1000);

export const VirtuosoExample: Story<Partial<VirtuosoProps>> = ({
  children,
  ...args
}) => {
  return (
    <>
      <Virtuoso
        style={{ height: '200px' }}
        totalCount={1000}
        data={menus}
        itemContent={(index) => {
          const label = `${index}-MenuItem`;
          return (
            <RcMenuItem key={index} data-search-text={label}>
              <RcListItemText primary={label} />
            </RcMenuItem>
          );
        }}
        {...args}
      />
    </>
  );
};

VirtuosoExample.args = {};

VirtuosoExample.argTypes = {
  ...notControlInDocTable<keyof VirtuosoProps>([]),
};

VirtuosoExample.parameters = {
  tags: [
    {
      name: 'Mui',
      href: 'https://material-ui.com/components/menus/',
      value: 'Menu',
    },
    {
      name: 'Source',
      value: 'virtuoso',
      href: 'https://virtuoso.dev/',
      color: 'avatar.lake',
    },
    {
      name: 'Accessibility',
      value: '100%',
    },
  ],
};
