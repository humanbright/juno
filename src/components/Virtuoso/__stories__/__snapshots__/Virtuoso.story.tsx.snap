// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸš€ Cleanup Components/Virtuoso Virtuoso Example 1`] = `
<div
  style={
    Object {
      "WebkitOverflowScrolling": "touch",
      "height": "200px",
      "outline": "none",
      "overflowY": "auto",
      "position": "relative",
    }
  }
  tabIndex={0}
>
  <div
    style={
      Object {
        "height": "100%",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "marginTop": 0,
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    />
  </div>
</div>
`;
